File: tools/office-link/auth_lock.py
"""Hikvision-style password attempt lock (app-side).

Wrong password → warn; after max_fails → short lock (operator can reset).
Network timeout / offline is not a password fail.
"""
from __future__ import annotations

from typing import Callable

CONFIRM = "confirm"
LOCKED = "locked"
IDLE = "idle"
SUCCESS = "success"

# Office use: enough tries, short lock (was 2 fails / 30 min — felt like “password always wrong”).
DEFAULT_LOCK_SECONDS = 3 * 60
DEFAULT_MAX_FAILS = 5


class AuthLock:
    def __init__(
        self,
        lock_seconds: int = DEFAULT_LOCK_SECONDS,
        now: Callable[[], float] | None = None,
        max_fails: int = DEFAULT_MAX_FAILS,
    ) -> None:
        self.lock_seconds = int(lock_seconds)
        self.max_fails = int(max_fails)
        self._now = now or __import__("time").time
        self.fail_count = 0
        self.lock_until = 0.0

    def _expire_if_due(self) -> None:
        if self.fail_count >= self.max_fails and self._now() >= self.lock_until:
            self.fail_count = 0
            self.lock_until = 0.0

    def is_locked(self) -> bool:
        self._expire_if_due()
        return self.fail_count >= self.max_fails and self._now() < self.lock_until

    def remaining_seconds(self) -> int:
        if not self.is_locked():
            return 0
        return max(0, int(self.lock_until - self._now()))

    def can_attempt(self) -> bool:
        return not self.is_locked()

    def phase(self) -> str:
        if self.is_locked():
            return LOCKED
        if self.fail_count >= 1:
            return CONFIRM
        return IDLE

    def record_timeout(self) -> None:
        """Network timeout is NOT a password fail."""
        return None

    def record_offline(self) -> None:
        """Device unreachable is NOT a password fail."""
        return None

    def record_401(self) -> str:
        if self.is_locked():
            return LOCKED
        self.fail_count += 1
        if self.fail_count >= self.max_fails:
            self.lock_until = self._now() + self.lock_seconds
            return LOCKED
        return CONFIRM

    def record_success(self) -> None:
        self.fail_count = 0
        self.lock_until = 0.0

    def reset(self) -> None:
        """Operator unlock — lock is app-side only, not on the terminal."""
        self.fail_count = 0
        self.lock_until = 0.0

    def format_remaining(self) -> str:
        sec = self.remaining_seconds()
        mins, rem = divmod(sec, 60)
        return f"{mins:02d}:{rem:02d}"

File: apps/office-link-mobile/lib/core/security/auth_lock.dart
/// Hikvision-style password attempt lock (app-side).
///
/// 1st 401 → confirm (re-type, no auto-retry).
/// 2nd 401 → lock for [lockSeconds] (default 30 min).
/// Network timeout / offline is not a password fail.
class AuthLock {
  AuthLock({
    this.lockSeconds = defaultLockSeconds,
    this.maxFails = 2,
    DateTime Function()? now,
  }) : _now = now ?? DateTime.now;

  static const int defaultLockSeconds = 30 * 60;
  static const String confirm = 'confirm';
  static const String locked = 'locked';
  static const String idle = 'idle';
  static const String success = 'success';

  final int lockSeconds;
  final int maxFails;
  final DateTime Function() _now;

  int failCount = 0;
  DateTime? lockUntil;

  void _expireIfDue() {
    if (failCount >= maxFails &&
        lockUntil != null &&
        !_now().isBefore(lockUntil!)) {
      failCount = 0;
      lockUntil = null;
    }
  }

  bool get isLocked {
    _expireIfDue();
    return failCount >= maxFails &&
        lockUntil != null &&
        _now().isBefore(lockUntil!);
  }

  int get remainingSeconds {
    if (!isLocked || lockUntil == null) return 0;
    final left = lockUntil!.difference(_now()).inSeconds;
    return left < 0 ? 0 : left;
  }

  bool get canAttempt => !isLocked;

  String get phase {
    if (isLocked) return locked;
    if (failCount == 1) return confirm;
    return idle;
  }

  void recordTimeout() {}
  void recordOffline() {}

  String record401() {
    if (isLocked) return locked;
    failCount += 1;
    if (failCount >= maxFails) {
      lockUntil = _now().add(Duration(seconds: lockSeconds));
      return locked;
    }
    return confirm;
  }

  void recordSuccess() {
    failCount = 0;
    lockUntil = null;
  }

  String formatRemaining() {
    final sec = remainingSeconds;
    final mins = sec ~/ 60;
    final rem = sec % 60;
    return '${mins.toString().padLeft(2, '0')}:${rem.toString().padLeft(2, '0')}';
  }
}
